File: tools/tex3dt.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-3.0-or-later
"""Decode the .3dt texture files that sit beside each .3DO.

Container and codec follow the notes in Omikron.txt. One .3dt holds every
texture of the matching .3DO, in material order; each is a palette followed by
image data:

    palette      3 bytes per colour, 16 colours if bpp == 4 else 256
    image data   imageDataSize bytes from the material record

`imageDataSize == 65536` means the 256x256 image is stored raw. Otherwise it is
an LZ scheme: a control byte supplies 8 flags, MSB first; a clear flag copies
one literal byte, a set flag introduces a two-or-more byte sequence token
giving a run length and a back-reference.

Container sizing is verified - for all 635 .3DO/.3dt pairs in gamedata/MESHES the
file length equals sum(paletteSize*3 + imageDataSize) over the materials.

The codec is SOLVED: every one of the **2534** textures under `gamedata/MESHES`
decodes to exactly width*height indices (`verify.py: textures`, --slow). The
parameter space the notes left open - run-length base, repeat-run count, the
bytes each token type consumes, the flag bit order and polarity, the token bit
packing - is settled by that, since a wrong choice in any of them desynchronises
and lands short or long.

`exact` is therefore True on everything shipped. It is still returned, and the
trim/pad below still runs, because `textures_bytes` is also used on the .3DO
files EMBEDDED in the SCX stream (the effect sprites, sprite_fx.py) and on
anything a caller hands it; a decoder that silently returned a short buffer
would be worse than one that says so.

    (An earlier version of this note said the decoder was a best fit of 96
    combinations landing exactly on about a third of textures. That was true
    when it was written and has not been true for some time - corrected
    2026-08-30, against the check that had been asserting 2534/2534 all along.)
"""
import struct, sys, os, zlib

def decode_image(data, want):
    """LZ-decompress `data` until `want` pixel indices are produced.

    Always returns exactly `want` bytes: a run that overshoots is trimmed, and
    a stream that ends early is padded by repeating the last pixel rather than
    leaving a black band. See the note in the module docstring about why some
    streams end early."""
    out = bytearray()
    i, n = 0, len(data)
    if n:
        out.append(data[0]); i = 1        # the first byte is always a literal
    while i < n and len(out) < want:
        group = data[i]; i += 1
        for bit in range(8):
            if len(out) >= want or i >= n: break
            if (group << bit) & 0x80 == 0:
                out.append(data[i]); i += 1
                continue
            token = data[i]; i += 1
            size = (token & 0xFC) // 4 + 3
            kind = token & 3
            if kind == 0:                       # repeat the previous pixel
                if out:
                    out += bytes([out[-1]]) * (size - 1)
            elif kind == 1:                     # back-reference, 8-bit offset
                off = data[i] + 1; i += 1
                for _ in range(size):
                    out.append(out[-off] if off <= len(out) else 0)
            elif kind == 2:                     # back-reference, 16-bit offset
                off = data[i] * 256 + data[i+1] + 1; i += 2
                for _ in range(size):
                    out.append(out[-off] if off <= len(out) else 0)
            else:                               # back-reference, offset * 256
                off = 256 * data[i]; i += 1
                for _ in range(size):
                    out.append(out[-off] if 0 < off <= len(out) else 0)
    exact = len(out) == want
    if len(out) > want:
        del out[want:]
    elif len(out) < want:
        out += bytes([out[-1] if out else 0]) * (want - len(out))
    return bytes(out), exact

def textures(dopath):
    """-> list of {name, w, h, bpp, rgb} for the model's materials."""
    d = open(dopath, "rb").read()
    tpath = None
    for cand in (dopath[:-4] + ".3dt", dopath[:-4] + ".3DT"):
        if os.path.exists(cand): tpath = cand; break
    if not tpath: return []
    return textures_bytes(d, open(tpath, "rb").read())


def textures_bytes(d, t):
    """The same, from bytes - the SCX effect sprites embed both (sprite_fx.py).

    The walk over `t` is exact: it consumes palette + image data per material
    and must land on len(t). `consumed` in the result of the last entry lets a
    caller assert that."""
    import mesh3do
    h = mesh3do.header_bytes(d, len(d))
    out, off = [], 0
    for i in range(h["materials"]):
        o = h["matOff"] + 80 * i
        name = d[o:o+20].split(b"\0")[0].decode("cp1252", "replace")
        # +60 data size, then the two RUNTIME slot fields (-1 in every shipped
        # material; the loader stamps the texture slot into +64 and the palette
        # slot into +68, and +64 is the low six bits of the render bucket key -
        # ASSETS 4b), then bits per pixel.
        size, _texSlot, _palSlot, bpp = struct.unpack_from("<4i", d, o + 60)
        w, ht = struct.unpack_from("<2h", d, o + 76)
        ncol = 16 if bpp == 4 else 256
        pal = t[off:off + ncol * 3]; off += ncol * 3
        raw = t[off:off + size];     off += size
        if size == 65536:
            idx, exact = raw[:w*ht].ljust(w*ht, b"\0"), len(raw) >= w * ht
        else:
            idx, exact = decode_image(raw, w * ht)
        rgb = bytearray(w * ht * 3)
        for k in range(min(len(idx), w * ht)):
            c = idx[k] * 3
            if c + 2 < len(pal):
                rgb[k*3:k*3+3] = pal[c:c+3]
        out.append({"name": name, "w": w, "h": ht, "bpp": bpp,
                    "exact": exact, "rgb": bytes(rgb), "consumed": off})
    return out

def png(w, h, rgb):
    def chunk(tag, data):
        return (struct.pack(">I", len(data)) + tag + data
                + struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF))
    raw = b"".join(b"\0" + rgb[y*w*3:(y+1)*w*3] for y in range(h))
    return (b"\x89PNG\r\n\x1a\n"
            + chunk(b"IHDR", struct.pack(">IIBBBBB", w, h, 8, 2, 0, 0, 0))
            + chunk(b"IDAT", zlib.compress(raw, 6))
            + chunk(b"IEND", b""))

if __name__ == "__main__":
    sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
    for p in sys.argv[1:]:
        for i, tx in enumerate(textures(p)):
            status = "exact" if tx["exact"] else "padded/trimmed"
            print(f"{os.path.basename(p)} [{i}] {tx['name']:14} "
                  f"{tx['w']}x{tx['h']} bpp={tx['bpp']} -> {status}")
            open(f"/tmp/{os.path.basename(p)[:-4]}_{i}.png", "wb").write(
                png(tx["w"], tx["h"], tx["rgb"]))

File: engine/src/formats/tex3dt.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
#include "formats/tex3dt.h"

#include "formats/mesh3do.h"

#if defined(__ARM_NEON) || defined(__ARM_NEON__)
#include <arm_neon.h>
#endif

namespace omk {

std::vector<std::uint8_t> decodeImage(std::span<const std::byte> data,
                                      std::size_t want, bool& exact) {
    std::vector<std::uint8_t> out;
    out.reserve(want);

    const std::size_t n = data.size();
    std::size_t i = 0;
    if (n != 0) {                       // the first byte is always a literal
        out.push_back(static_cast<std::uint8_t>(data[0]));
        i = 1;
    }

    // A back-reference may read a byte this run has just written, so the copy
    // is one at a time rather than a block move - runs that overlap their own
    // source are how the format expresses a repeating pattern.
    const auto copyBack = [&out](std::size_t off, int size) {
        for (int k = 0; k < size; ++k)
            out.push_back(off != 0 && off <= out.size()
                              ? out[out.size() - off]
                              : std::uint8_t{0});
    };

    while (i < n && out.size() < want) {
        const auto group = static_cast<std::uint8_t>(data[i++]);
        for (int bit = 0; bit < 8; ++bit) {
            if (out.size() >= want || i >= n) break;

            if (((group << bit) & 0x80) == 0) {          // a literal
                out.push_back(static_cast<std::uint8_t>(data[i++]));
                continue;
            }

            const auto token = static_cast<std::uint8_t>(data[i++]);
            const int  size  = (token & 0xFC) / 4 + 3;
            switch (token & 3) {
                case 0:                                  // repeat the last
                    if (!out.empty()) {
                        const auto last = out.back();
                        out.insert(out.end(), static_cast<std::size_t>(size - 1), last);
                    }
                    break;
                case 1: {                                // 8-bit offset
                    if (i >= n) { i = n; break; }
                    const std::size_t off = static_cast<std::uint8_t>(data[i++]) + 1u;
                    copyBack(off, size);
                    break;
                }
                case 2: {                                // 16-bit, big-endian
                    if (i + 1 >= n) { i = n; break; }
                    const std::size_t hi = static_cast<std::uint8_t>(data[i++]);
                    const std::size_t lo = static_cast<std::uint8_t>(data[i++]);
                    copyBack(hi * 256u + lo + 1u, size);
                    break;
                }
                default: {                               // offset * 256
                    if (i >= n) { i = n; break; }
                    const std::size_t off = 256u * static_cast<std::uint8_t>(data[i++]);
                    copyBack(off, size);
                    break;
                }
            }
        }
    }

    exact = out.size() == want;
    if (out.size() > want) {
        out.resize(want);
    } else if (out.size() < want) {
        const std::uint8_t pad = out.empty() ? std::uint8_t{0} : out.back();
        out.resize(want, pad);
    }
    return out;
}

std::vector<Texture> textures(std::span<const std::byte> d,
                              std::span<const std::byte> t) {
    std::vector<Texture> out;
    const auto header = readHeader(d);
    if (!header) return out;

    std::size_t off = 0;
    for (int i = 0; i < header->materials; ++i) {
        const auto mat = readMaterial(d, *header, i);
        if (!mat) break;

        Texture tx;
        tx.name   = mat->name;
        tx.width  = mat->width;
        tx.height = mat->height;
        tx.bpp    = mat->bpp;

        const std::size_t ncol = (mat->bpp == 4) ? 16u : 256u;
        const std::size_t palBytes = ncol * 3u;

        // The walk consumes palette then data and must land on t.size(); a
        // short tail means the pair does not belong together, so stop rather
        // than read past it.
        if (off + palBytes > t.size()) break;
        const auto pal = t.subspan(off, palBytes);
        off += palBytes;

        const auto want = static_cast<std::size_t>(mat->width) *
                          static_cast<std::size_t>(mat->height);
        const auto dataSize = static_cast<std::size_t>(mat->dataSize < 0 ? 0 : mat->dataSize);
        if (off + dataSize > t.size()) break;
        const auto raw = t.subspan(off, dataSize);
        off += dataSize;

        std::vector<std::uint8_t> idx;
        if (dataSize == 65536) {                 // stored raw, 256x256
            idx.assign(raw.size() >= want ? want : raw.size(), 0);
            for (std::size_t k = 0; k < idx.size(); ++k)
                idx[k] = static_cast<std::uint8_t>(raw[k]);
            tx.exact = raw.size() >= want;
            idx.resize(want, 0);
        } else {
            idx = decodeImage(raw, want, tx.exact);
        }

        tx.rgb.assign(want * 3u, 0);
        std::uint8_t* const px = tx.rgb.mutableData();   // one detach check, not one a texel
        for (std::size_t k = 0; k < want && k < idx.size(); ++k) {
            const std::size_t c = static_cast<std::size_t>(idx[k]) * 3u;
            if (c + 2 < pal.size()) {
                px[k * 3    ] = static_cast<std::uint8_t>(pal[c    ]);
                px[k * 3 + 1] = static_cast<std::uint8_t>(pal[c + 1]);
                px[k * 3 + 2] = static_cast<std::uint8_t>(pal[c + 2]);
            }
        }
        out.push_back(std::move(tx));
    }
    return out;
}

void rgbToRgbaKeyedGeneric(const std::uint8_t* rgb, std::uint8_t* rgba, std::size_t n) {
    for (std::size_t k = 0; k < n; ++k) {
        rgba[4 * k + 0] = rgb[3 * k + 0];
        rgba[4 * k + 1] = rgb[3 * k + 1];
        rgba[4 * k + 2] = rgb[3 * k + 2];
        rgba[4 * k + 3] = (rgb[3 * k] | rgb[3 * k + 1] | rgb[3 * k + 2]) ? 255 : 0;
    }
}

#if defined(__ARM_NEON) || defined(__ARM_NEON__)
void rgbToRgbaKeyedNeon(const std::uint8_t* rgb, std::uint8_t* rgba, std::size_t n) {
    std::size_t k = 0;
    const uint8x16_t zero = vdupq_n_u8(0);
    for (; k + 16 <= n; k += 16) {
        const uint8x16x3_t c = vld3q_u8(rgb + 3 * k);
        uint8x16x4_t o;
        o.val[0] = c.val[0];
        o.val[1] = c.val[1];
        o.val[2] = c.val[2];
        // any channel nonzero -> 0xFF, black -> 0
        o.val[3] = vcgtq_u8(vorrq_u8(vorrq_u8(c.val[0], c.val[1]), c.val[2]), zero);
        vst4q_u8(rgba + 4 * k, o);
    }
    if (k < n) rgbToRgbaKeyedGeneric(rgb + 3 * k, rgba + 4 * k, n - k);
}
#else
void rgbToRgbaKeyedNeon(const std::uint8_t* rgb, std::uint8_t* rgba, std::size_t n) {
    rgbToRgbaKeyedGeneric(rgb, rgba, n);
}
#endif

void rgbToRgbaKeyed(const std::uint8_t* rgb, std::uint8_t* rgba, std::size_t n) {
#if defined(__ARM_NEON) || defined(__ARM_NEON__)
    rgbToRgbaKeyedNeon(rgb, rgba, n);
#else
    rgbToRgbaKeyedGeneric(rgb, rgba, n);
#endif
}

}  // namespace omk
